updates_and_signals/atomic_message_handlers/workflow.py

```diff
@@ -0,0 +1,198 @@
+import asyncio
+import logging
+import uuid
+from dataclasses import dataclass
+from datetime import timedelta
+from typing import Dict, List, Optional
+
+from temporalio import activity, common, workflow
+from temporalio.client import Client, WorkflowHandle
+from temporalio.common import RetryPolicy
+from temporalio.worker import Worker
+
+from updates_and_signals.atomic_message_handlers.activities import (
+    AllocateNodesToJobInput,
+    DeallocateNodesForJobInput,
+    FindBadNodesInput,
+    allocate_nodes_to_job,
+    deallocate_nodes_for_job,
+    find_bad_nodes,
+)
+
+
+# In workflows that continue-as-new, it's convenient to store all your state in one serializable structure
+# to make it easier to pass between runs
+@dataclass(kw_only=True)
+class ClusterManagerState:
+    cluster_started: bool = False
+    cluster_shutdown: bool = False
+    nodes: Optional[Dict[str, Optional[str]]] = None
+    max_assigned_nodes: int = 0
+    num_assigned_nodes: int = 0
+
+
+@dataclass(kw_only=True)
+class ClusterManagerInput:
+    state: Optional[ClusterManagerState] = None
+    test_continue_as_new: bool = False
+
+
+@dataclass
+class ClusterManagerResult:
+    max_assigned_nodes: int
+    num_assigned_nodes: int
```

Comment:
Porting this to .NET and not sure there is value storing this field on "state" (and it's built out of lock, so it's a bit confusing)